### DISCUSSION TITLE
SPO strength

### DISCUSSION THREAD

**Discussion by DriptD** [mtex-toolbox/mtex#769] [Ask Anything] [answered]
Hello,

I have a set of two SPO rose plots produced using the **_area-weighted method_** suggested in the MTEX documentation. I am trying to quantify the strength of individual SPO plots using the formula of Barraud (2006, JVGR) (screenshot below). I'm using **_MTEX 5.6.0 in Matlab 2020b_**.

![image](https://user-images.githubusercontent.com/77139598/107737259-0aab8900-6d2a-11eb-96b6-b736abc767a6.png)

However, I'm not sure if I get the correct result when the no. of grains is <50. Below is the part of the code:

**********************************************************************************
grain_LA = grains('phase_name').longAxis; 
grain_length = grains('phase_name').diameter;

t_sum = [0 0; 0 0]; % empty square matrix to store the sum of **_'n'_** square matrices

x2 = (grain_LA.x).^2; % square of the x-component of the long axis unit vector

xy = (grain_LA.x).*(grain_LA.y); % product of the x- and y- components of the long axis unit vector

y2 = (grain_LA.y).^2; % square of the y-component of the long axis unit vector

for n = 1:length(grain_LA)
    t{1,n} = [x2(n,:),xy(n,:);xy(n,:),y2(n,:)].*grain_length(n,:);
    t_sum = t_sum + t{1,n};
end
D = t_sum./n;
e = eig(D)';
AF = (max(e)-min(e))/(max(e))*100;
**********************************************************************************

Here is what I got:

![image](https://user-images.githubusercontent.com/77139598/107737022-780aea00-6d29-11eb-971b-ab6000b789d5.png)

Following are the questions I have:
**1.** I think the plot to the right should have a lower AF because the grain set seems to have a less preferred alignment than that of the other. But it isn't so. Could anyone please point out if there's something wrong with the code? 

**2.** Is it that the SPOs are '_area-weighted_', and while computing the AF, I'm not taking that into account? Could anyone please suggest how to do that? 

**3.** Is there any other method that could be used to quantify the SPO strength instead of this one?

Thank you !!

Best,
Dripta

**Comment by kilir** [comment, marked as answer]
Hi Dripta,
you should probably treat your grain long axes as axes not directions - unless you have a way to distinguish between the "head" and the "tail" of a grain.

You should be able to weight your data by scaling the length of your grain long axes. Mtex also has `eig()` for vector3d.

Cheers,
Rüdiger

**Reply by DriptD** [reply]
Hello Rüdiger,

I'm sorry I think I didn't get you completely. 

I used the x and y components of the long axis vector instead of the angles (rho of the vector3d variable grain_LA). I guess the long axes are unit vectors such that (and please correct me if I'm wrong)
grain_LA.x = cos(grain_LA.rho) and grain_LA.y = sin(grain_LA.rho)

Scaling of the grain long axes lengths somewhat improved the result (the ratio of high to low AF decreased to 1.04 from 3.36 - see image below), however, I'm still unable to understand why the AF value for the set of grains to the right is higher !! 

![image](https://user-images.githubusercontent.com/77139598/107751336-4a329f00-6d43-11eb-9dd0-e9f562af04b3.png)


Thank you very much for the response !!

Best,
Dripta

**Reply by kilir** [reply]
Hi
you compute an eigenvalue from directed data and from that point of view, it seems obvious to me to expect the second (disregarding any weighting and so) dataset to show a higher degree of ordering.
Can you really distinguish polarity of a grain long axis? Mtex usuall does not and the vector returned by longAxis is antipodal. If you'd stick with this and `vector3d/eig()`, you should be fine. If you want to compute those yourself, you have to consider the directionality.

In general, it's a good idea to text those things on an artificial dataset such that you know the result you should get.

Cheers,
Rüdiger

**Reply by DriptD** [reply]
Hello Rüdiger

Thank you for the response. I think I need some time to get my head around this thing and code it. I really appreciate your help.

Regards,
Dripta

**Reply by kilir** [reply]
Mtex does this already.

![testellipses](https://user-images.githubusercontent.com/11697557/107857277-c86b7000-6e2d-11eb-8d11-08293543f7ac.png)
```

% use test ellipses
plotx2east
plotzOutOfPlane

lax = grains.longAxis;
plot(grains,lax.rho/degree)
mtexColorMap(colorcet('C2'))
mtexColorbar

subplot(1,3,1)
% frequency histogram of long axes
histogram(lax,36)

[ev,e] = eig(lax);
mtexTitle(['e1: ' num2str(round(100*e(3)/sum(e))/100) ...
          ' at '   num2str(round(10*mod(ev(3).rho,pi)/degree)/10) ])

subplot(1,3,2)
% weight by length
histogram(lax, 'weights',grains.diameter,36)

[ev,e] = eig(lax.*grains.diameter);
mtexTitle(['length weighted e1: ' num2str(round(100*e(3)/sum(e))/100) ...
          ' at '   num2str(round(10*mod(ev(3).rho,pi)/degree)/10) ])

subplot(1,3,3)
% weight by area
histogram(lax, 'weights',grains.area,36)

[ev,e] = eig(lax.*grains.area);
mtexTitle(['area weighted e1: ' num2str(round(100*e(3)/sum(e))/100) ...
          ' at '   num2str(round(10*mod(ev(3).rho,pi)/degree)/10) ])

```
![testelli_histo](https://user-images.githubusercontent.com/11697557/107857279-cbfef700-6e2d-11eb-8202-897e530a1e4c.png)

Cheers,
Rüdiger

**Reply by DriptD** [reply]
Hello Rüdiger,

Thank you so much for the elaborate illustration. From what I see (and got to know from reading on eigen analysis), I suppose the largest eigenvalue approaches 1 as the distribution becomes more uniform/unimodal/ordered. Am I right?

This approach worked well for the data I have wherein the no. of grains is low (<100). However, for a larger set of relatively undeformed quartz grains (>20k) the largest eigenvalues are near close even when the SPO shows otherwise. Here is the output (area-weighted data):

![test1](https://user-images.githubusercontent.com/77139598/108026807-11e6d580-704f-11eb-80e8-04fcf667c075.jpg)

I'm not sure yet what could be the reason for this !!

Thanks again.

Best,
Dripta

**Reply by kilir** [reply]
Hi,
a) largest eigenvalue approaches 1 for a delta /single value distribution.
b) How do you compute your e1 value? It may look like, you do treat your axes as undirected (despite the plot looks like). Code might help.
Cheers,
Rüdiger

**Reply by DriptD** [reply]
Hello,

I used the code you posted earlier in this thread.

grain_LA = grains('Quartz').longAxis;
grain_area = grains('Quartz').area;

[ev,e] = eig(grain_LA.*grain_area); ------------------------------------ (i)
e1 = e(3)/sum(e);

The weights used to plot the SPO were rather calculated in the following way:
weights = grains('Quartz').area .* (grains('Quartz').aspectRatio-1);  ----------------------------- (ii)

Out of curiosity, I also used the weights from (ii) in (i) to check if scaling by the area could be the error. But even then the e1 value for the second plot came out to be lower than the other.

Best,
Dripta

**Reply by kilir** [reply]
Hi,
I'd suggest plotting the distribution you compute the eigenvalues of or the other way around, not two different things. Where do your mean directions end up?
Cheers Rüdiger

**Reply by DriptD** [reply]
Hello,

The mean directions 14.6 and 53

![test1](https://user-images.githubusercontent.com/77139598/108047139-77e15600-706b-11eb-8f67-2f6b295c896e.jpg)

Best,
Dripta

**Reply by kilir** [reply]
53 seems a little odd.
The histograms you show are plotted with the same weights you use compute the eigenvalues/vectors?

Cheers,
Rüdiger

**Reply by DriptD** [reply]
Oh, I'm so sorry.

If the same weights are used for both the histogram and eigenvalue calculation, then the mean directions are 0.1 and 38.8 for the first and second plots, respectively.

The e1 values are 0.65 and 0.54 in that case.

Best,
Dripta

**Reply by kilir** [reply]
I still don't really see the 38 degree, so something seems to be not right. If you like, you could share the variables grain_LA and grain_area. (Just save them as mat and zip them up)
Cheers,
Rüdiger

**Reply by DriptD** [reply]
Hello,

Here it is. 
[Grain_LA_Area.zip](https://github.com/mtex-toolbox/mtex/files/5988042/Grain_LA_Area.zip)


Best,
Dripta

**Reply by kilir** [reply]
I think you have a sampling problem with the data. Smoothing grains may help, but in general, you should discard grains below a certain size (i.e. < 50 pixels, but depends on the type of analysis) for shape analysis.

First figure: your data as is
<img width="668" alt="dd1_raw" src="https://user-images.githubusercontent.com/11697557/108058115-d569a880-7053-11eb-8ef1-d196a339ac03.png">

Second figure: only those direction corresponding to areas > 12 (wild guess, no idea what this is in pixels in your case)
<img width="668" alt="dd1_clean" src="https://user-images.githubusercontent.com/11697557/108058222-021dc000-7054-11eb-950a-9d3d24e954b9.png">

And here, the reasonable data as density:
<img width="513" alt="dd1_den" src="https://user-images.githubusercontent.com/11697557/108059213-75740180-7055-11eb-8c80-8b84e255d1a7.png">


Cheers,
Rüdiger

**Reply by DriptD** [reply]
Hello,

Thanks a lot. I was thinking of discarding finer grains and during processing the raw data I had only considered grains with at least 3 pixels. But I guess, maybe for the shape analysis, I'll have to raise that cutoff.

Thanks again.

Best,
Dripta
